Sources/Networking/v2/APIService.swift

```diff
@@ -0,0 +1,126 @@
+//
+//  APIService.swift
+//
+//  Copyright © 2024 DuckDuckGo. All rights reserved.
+//
+//  Licensed under the Apache License, Version 2.0 (the "License");
+//  you may not use this file except in compliance with the License.
+//  You may obtain a copy of the License at
+//
+//  http://www.apache.org/licenses/LICENSE-2.0
+//
+//  Unless required by applicable law or agreed to in writing, software
+//  distributed under the License is distributed on an "AS IS" BASIS,
+//  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//  See the License for the specific language governing permissions and
+//  limitations under the License.
+//
+
+import Foundation
+import os.log
+
+public protocol APIService {
+
+    typealias APIResponse = (data: Data?, httpResponse: HTTPURLResponse)
+
+    func fetch<T: Decodable>(request: APIRequestV2) async throws -> T?
+    func fetch(request: APIRequestV2) async throws -> APIService.APIResponse
+}
+
+public struct DefaultAPIService: APIService {
+    private let urlSession: URLSession
+
+    public init(urlSession: URLSession = .shared) {
+        self.urlSession = urlSession
+
+    }
+
+    /// Fetch an API Request
+    /// - Parameter request: A configured APIRequest
+    /// - Returns: An instance of the inferred decodable object, can be a `String` or any `Decodable` model, nil if the response body is empty
+    public func fetch<T: Decodable>(request: APIRequestV2) async throws -> T? {
+        let response: APIService.APIResponse = try await fetch(request: request)
+
+        guard let data = response.data else {
+            return nil
+        }
+
+        try Task.checkCancellation()
+
+        // Try to decode the data
+        Logger.networking.debug("Decoding response body as \(T.self)")
+        switch T.self {
+        case is String.Type:
+            guard let resultString = String(data: data, encoding: .utf8) as? T else {
+                let error = APIRequestV2.Error.invalidDataType
+                Logger.networking.error("Error: \(error.localizedDescription)")
+                throw error
+            }
+            return resultString
+        default:
+            // Decode data
+            let decoder = JSONDecoder()
+            return try decoder.decode(T.self, from: data)
```

Comment:
Although we are switching on the type (strong -vs- other), we are really just decoding in both cases. So, maybe we could extract all of this to a `decode` method. This method could throw an error which we then handle and throw as `APIRequestV2.Error.invalidDataType`. This would mean we are consistent in throwing `APIRequestV2.Error.invalidDataType` when decoding fails for either string or other types. e.g:

```
  private func decode<T: Decodable>(data: Data, to type: T.Type) throws -> T {
        if T.self == String.self, let resultString = String(data: data, encoding: .utf8) as? T {
            return resultString
        } else {
            return try decoder.decode(T.self, from: data)
        }
    }
```

and then the callsite is 

```
 do {
            return try decode(data: data, to: T.self)
        } catch {
            Logger.networking.error("Error decoding data: \(error.localizedDescription)")
            throw APIRequestV2.Error.invalidDataType
        }
```

Reply:
This and the previous comments are on point, I'm actually thinking of removing this call completely and leaving the decoding to the user. I added this call as a utility but I have never used it, I always need the HTTPURLResponse for checking response codes and headers.

Reply:
Your call. If we don’t include decoding now we won’t be prevented from adding it later.

Reply:
So, I changed the approach completely, turns out that doing slightly more advanced stuff with tuples is a pain, so:

1. I created a new `APIResponseV2` struct and removed the APIResponse tuple, the content is the same
2. I removed `fetch<T: Decodable>(request: APIRequestV2) async throws -> T?`
3. I created a `APIResponseV2` extension with a function `func decodeBody<T: Decodable>(decoder: JSONDecoder = JSONDecoder()) throws -> T` use:

```
let response = try await apiService.fetch(request: request)
let responseHTML: String = try response.decodeBody()
```
Much more clean and simple + the decoder can be injected, as suggested